### ISSUE TITLE
Metal: search selects LLIR graphs that produce NaN logits nondeterministically

### ISSUE THREAD

**Issue by BABTUNA** [jafioti/luminal#410]
# Metal: search selects LLIR graphs that produce NaN logits nondeterministically

## Summary

Running a new (unmerged) Phi-4-mini example on the Metal backend, identical
inputs and identical config produce different numeric results across runs.
Some runs return correct logits, some return all-NaN, and at least one run
returned *exactly half* the vocab as NaN.

This appears to violate the core contract in `spec.md`:

> Every selectable LLIR graph for a given HLIR graph must be semantically
> equivalent for all valid inputs, dtypes, shapes, strides, and dynamic dims.

## Observed

Same graph, same prompt, `search_graph_limit(10)`, 4 layers, F32 weights:

```
run 1:      0 / 200064 logits finite   (all NaN)
run 2: 200064 / 200064 logits finite   (correct)
run 3:      0 / 200064 logits finite   (all NaN)
run 4: 100032 / 200064 logits finite   (exactly half NaN)
```

The half-NaN run is the interesting one: 100032 = 200064 / 2 exactly. That
looks structural (a partitioned/tiled output where one partition is wrong)
rather than numerical.

Also reproduced at F16, so it is not dtype range exhaustion:

```
F16, 4 layers:  NaN, NaN, finite
```

Varying layer count is not monotonic either (2 finite, 4 NaN, 8 finite,
16 finite, 24 NaN, 32 NaN), consistent with per-run search selection rather
than depth.

## Environment

- M1 Pro (14-core GPU), 16GB, macOS 25.4
- `luminal` @ 62d3cc0d, Metal backend
- rustc 1.97.1

## Caveats

- The reproducer is currently a ~1000-line example crate, not minimized. Happy
  to cut it down if that is useful.
- The model is the first example to slice fused QKV / gate_up projections, so
  it feeds non-contiguous slice views into downstream matmuls in a way the
  existing examples do not. That may be the trigger.
- Vocab is 200064, larger than any current example (Qwen 151936, Llama 128256).

## Question

Is this a known issue, and is a minimized repro against `ReferenceRuntime`
the most useful next step?
